File: src/ios/TrackierCordovaPlugin.swift
import Cordova
import trackier_ios_sdk

@objc(TrackierCordovaPlugin) 
class TrackierCordovaPlugin : CDVPlugin {
    @objc(initializeSDK:)
    func initializeSDK(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        let msg = command.arguments[0] as? String ?? ""
        if (msg.count > 0){
            let dict = convertStringToDictionary(text: msg)
            let appToken = dict?["appToken"] as! String;
            let environment = dict?["environment"] as! String;
            let config = TrackierSDKConfig(appToken: appToken , env: environment)
            config.setSDKType(sdkType: "cordova_sdk")
            config.setSDKVersion(sdkVersion: "1.6.48")
            TrackierSDK.initialize(config: config)

            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: msg
            )
        }
        self.commandDelegate!.send(
            pluginResult, 
            callbackId: command.callbackId
        )
    }

    @objc(setUserId:)
    func setUserId(command: CDVInvokedUrlCommand){
        let msg = command.arguments[0] as? String ?? ""
        TrackierSDK.setUserID(userId: msg)
    }

    @objc(setUserEmail:)
    func setUserEmail(command: CDVInvokedUrlCommand){
        let msg = command.arguments[0] as? String ?? ""
        TrackierSDK.setUserEmail(userEmail: msg)
    }

    @objc(setUserName:)
    func setUserName(command: CDVInvokedUrlCommand){
        let msg = command.arguments[0] as? String ?? ""
        TrackierSDK.setUserName(userName: msg)
    }

    @objc(setUserPhone:)
    func setUserPhone(command: CDVInvokedUrlCommand){
        let msg = command.arguments[0] as? String ?? ""
        TrackierSDK.setUserPhone(userPhone: msg)
    }

    @objc(setDOB:)
    func setDOB(command: CDVInvokedUrlCommand){
        let msg = command.arguments[0] as? String ?? ""
        TrackierSDK.setDOB(dob: msg)
    }
    
    @objc(setGender:)
    func setGender(command: CDVInvokedUrlCommand){
        let msg = command.arguments[0] as? String ?? ""
        switch msg {
            case "Male": TrackierSDK.setGender(gender: .MALE)
            case "Female": TrackierSDK.setGender(gender: .FEMALE)
            case "Others": TrackierSDK.setGender(gender: .OTHERS)
            default: print("No Gender found")
        }
    }

    @objc(getTrackierId:)
    func getTrackierId(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getTrackierId())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getAd:)
    func getAd(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getAd())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getAdID:)
    func getAdID(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getAdID())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getCampaign:)
    func getCampaign(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getCampaign())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getCampaignID:)
    func getCampaignID(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getCampaignID())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getAdSetID:)
    func getAdSetID(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getAdSetID())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getChannel:)
    func getChannel(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getChannel())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getP1:)
    func getP1(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getP1())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getP2:)
    func getP2(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getP2())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getP3:)
    func getP3(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getP3())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getP4:)
    func getP4(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getP4())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getP5:)
    func getP5(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getP5())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getClickId:)
    func getClickId(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getClickId())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getDlv:)
    func getDlv(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getDlv())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getPid:)
    func getPid(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getPid())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getIsRetargeting:)
    func getIsRetargeting(command: CDVInvokedUrlCommand){
        var pluginResult: CDVPluginResult?
        pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TrackierSDK.getIsRetargeting())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    @objc(trackEvent:)
    func trackEvent(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )
        let msg = command.arguments[0] as? String ?? ""
        if (msg.count > 0) {
            let dict = convertStringToDictionary(text: msg)
            let orderId: String = dict?["orderId"] as? String ?? ""
            let currency: String = dict?["currency"] as? String ?? ""
            let revenue: Float64 = dict?["revenue"] as? Float64 ?? 0.0
            let discount: Float64 = dict?["discount"] as? Float64 ?? 0.0
            let couponCode: String = dict?["couponCode"] as? String ?? ""
            let eventId: String = dict?["eventId"] as! String
            let param1: String = dict?["param1"] as? String ?? ""
            let param2: String = dict?["param2"] as? String ?? ""
            let param3: String = dict?["param3"] as? String ?? ""
            let param4: String = dict?["param4"] as? String ?? ""
            let param5: String = dict?["param5"] as? String ?? ""
            let param6: String = dict?["param6"] as? String ?? ""
            let param7: String = dict?["param7"] as? String ?? ""
            let param8: String = dict?["param8"] as? String ?? ""
            let param9: String = dict?["param9"] as? String ?? ""
            let param10: String = dict?["param10"] as? String ?? ""
            var ev:Dictionary<String,Any> = dict?["ev"] as? Dictionary<String,Any> ?? [:]
            for (key, value) in ev {
                ev[key] = value
            }
            let event = TrackierEvent(id: eventId)
            event.setRevenue(revenue: revenue, currency: currency)
            event.orderId = orderId
            event.setCouponCode(couponCode: couponCode)
            event.setDiscount(discount: discount)
            event.param1  = param1
            event.param2  = param2
            event.param3  = param3
            event.param4  = param4
            event.param5  = param5
            event.param6  = param6
            event.param7  = param7
            event.param8  = param8
            event.param9  = param9
            event.param10 = param10
            for (key, value) in ev {
                event.addEventValue(prop: key, val: value)
            }
            TrackierSDK.trackEvent(event: event)

            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: msg
            )
        }
        self.commandDelegate!.send(
            pluginResult, 
            callbackId: command.callbackId
        )

    }

    func convertStringToDictionary(text: String) -> [String:AnyObject]? {
        if let data = text.data(using: .utf8) {
            do {
                let json = try JSONSerialization.jsonObject(with: data, options: .mutableContainers) as? [String:AnyObject]
                return json
            } catch {}
        }
        return nil
    }
}

File: src/android/TrackierCordovaPlugin.java
package com.trackier.cordova_sdk;

import android.net.Uri;
import android.util.Log;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONObject;
import com.trackier.cordova_sdk.TrackierCordovaUtil;
import com.trackier.sdk.TrackierSDK;

/**
 * This class echoes a string called from JavaScript.
 */
public class TrackierCordovaPlugin extends CordovaPlugin {

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) {
        try {
            if (action.equals("initializeSDK")) {
                String message = args.getString(0);
                this.initializeSDK(message, callbackContext);
                return true;
            } else if (action.equals("trackEvent")) {
                String message = args.getString(0);
                this.trackEvent(message, callbackContext);
                return true;
            } else if (action.equals("setUserId")) {
                return setUserId(TrackierCordovaUtil.optString(args, 0));
            } else if (action.equals("setUserEmail")) {
                return setUserEmail(TrackierCordovaUtil.optString(args, 0));
            } else if (action.equals("setUserName")) {
                return setUserName(TrackierCordovaUtil.optString(args, 0));
            } else if (action.equals("setUserPhone")) {
                return setUserPhone(TrackierCordovaUtil.optString(args, 0));
            } else if (action.equals("setDOB")) {
                return setDOB(TrackierCordovaUtil.optString(args, 0));
            } else if (action.equals("setGender")) {
                return setGender(TrackierCordovaUtil.optString(args, 0));
            } else if (action.equals("getTrackierId")) {
                String trackierId = com.trackier.sdk.TrackierSDK.getTrackierId();
                callbackContext.success(trackierId);
            } else if (action.equals("getAd")) {
                String data = com.trackier.sdk.TrackierSDK.getAd();
                callbackContext.success(data);
            } else if (action.equals("getAdID")) {
                String data = com.trackier.sdk.TrackierSDK.getAdID();
                callbackContext.success(data);
            } else if (action.equals("getCampaign")) {
                String data = com.trackier.sdk.TrackierSDK.getCampaign();
                callbackContext.success(data);
            } else if (action.equals("getCampaignID")) {
                String data = com.trackier.sdk.TrackierSDK.getCampaignID();
                callbackContext.success(data);
            } else if (action.equals("getAdSet")) {
                String data = com.trackier.sdk.TrackierSDK.getAdSet();
                callbackContext.success(data);
            } else if (action.equals("getAdSetID")) {
                String data = com.trackier.sdk.TrackierSDK.getAdSetID();
                callbackContext.success(data);
            } else if (action.equals("getChannel")) {
                String data = com.trackier.sdk.TrackierSDK.getChannel();
                callbackContext.success(data);
            } else if (action.equals("getP1")) {
                String data = com.trackier.sdk.TrackierSDK.getP1();
                callbackContext.success(data);
            } else if (action.equals("getP2")) {
                String data = com.trackier.sdk.TrackierSDK.getP2();
                callbackContext.success(data);
            } else if (action.equals("getP3")) {
                String data = com.trackier.sdk.TrackierSDK.getP3();
                callbackContext.success(data);
            } else if (action.equals("getP4")) {
                String data = com.trackier.sdk.TrackierSDK.getP4();
                callbackContext.success(data);
            } else if (action.equals("getP5")) {
                String data = com.trackier.sdk.TrackierSDK.getP5();
                callbackContext.success(data);
            } else if (action.equals("getClickId")) {
                String data = com.trackier.sdk.TrackierSDK.getClickId();
                callbackContext.success(data);
            } else if (action.equals("getDlv")) {
                String data = com.trackier.sdk.TrackierSDK.getDlv();
                callbackContext.success(data);
            } else if (action.equals("getPid")) {
                String data = com.trackier.sdk.TrackierSDK.getPid();
                callbackContext.success(data);
            } else if (action.equals("getIsRetargeting")) {
                String data = com.trackier.sdk.TrackierSDK.getIsRetargeting();
                callbackContext.success(data);
            }
        } catch (Exception e) {

        }
        return false;
    }

    private void initializeSDK(String message, CallbackContext callbackContext) {
        if (message == null && message.length() == 0) {
            callbackContext.error("Expected one non-empty string argument.");
            return;
        }
        try {
            JSONObject trackiersdkConfigJson = new JSONObject(message);
            com.trackier.sdk.TrackierSDKConfig sdkConfig = new com.trackier.sdk.TrackierSDKConfig(webView.getContext(),
            TrackierCordovaUtil.getStringVal("appToken", trackiersdkConfigJson),
            TrackierCordovaUtil.getStringVal("environment", trackiersdkConfigJson));
            sdkConfig.setAppSecret(TrackierCordovaUtil.getStringVal("secretId", trackiersdkConfigJson), TrackierCordovaUtil.getStringVal("secretKey", trackiersdkConfigJson));
            sdkConfig.setManualMode(TrackierCordovaUtil.getBooleanVal("manualMode", trackiersdkConfigJson));
            sdkConfig.disableOrganicTracking(TrackierCordovaUtil.getBooleanVal("disableorganic", trackiersdkConfigJson));
            sdkConfig.setSDKType("cordova_sdk");
            sdkConfig.setSDKVersion("1.6.56");
            com.trackier.sdk.TrackierSDK.initialize(sdkConfig);
        } catch (Exception exception) {

        }

        callbackContext.success(message);
    }

    private void trackEvent(String message, CallbackContext callbackContext) {
        if (message == null || message.length() == 0) {
            callbackContext.error("Expected one non-empty string argument.");
            return;
        }
        try {
            JSONObject trackierEventJson = new JSONObject(message);
            com.trackier.sdk.TrackierEvent trackierEvent = new com.trackier.sdk.TrackierEvent(
            TrackierCordovaUtil.getStringVal("eventId", trackierEventJson));
            trackierEvent.orderId = TrackierCordovaUtil.getStringVal("orderId", trackierEventJson);
            trackierEvent.currency = TrackierCordovaUtil.getStringVal("currency", trackierEventJson);
            trackierEvent.discount = TrackierCordovaUtil.getFloatVal("discount", trackierEventJson);
            trackierEvent.couponCode = TrackierCordovaUtil.getStringVal("couponCode", trackierEventJson);
            trackierEvent.param1 = TrackierCordovaUtil.getStringVal("param1", trackierEventJson);
            trackierEvent.param2 = TrackierCordovaUtil.getStringVal("param2", trackierEventJson);
            trackierEvent.param3 = TrackierCordovaUtil.getStringVal("param3", trackierEventJson);
            trackierEvent.param4 = TrackierCordovaUtil.getStringVal("param4", trackierEventJson);
            trackierEvent.param5 = TrackierCordovaUtil.getStringVal("param5", trackierEventJson);
            trackierEvent.param6 = TrackierCordovaUtil.getStringVal("param6", trackierEventJson);
            trackierEvent.param7 = TrackierCordovaUtil.getStringVal("param7", trackierEventJson);
            trackierEvent.param8 = TrackierCordovaUtil.getStringVal("param8", trackierEventJson);
            trackierEvent.param9 = TrackierCordovaUtil.getStringVal("param9", trackierEventJson);
            trackierEvent.param10 = TrackierCordovaUtil.getStringVal("param10", trackierEventJson);
            trackierEvent.revenue = TrackierCordovaUtil.getDoubleVal("revenue", trackierEventJson);
            com.trackier.sdk.TrackierSDK.trackEvent(trackierEvent);

        } catch (Exception exception) {

        }

        callbackContext.success(message);
    }

    private boolean setUserId(String optString) {
        com.trackier.sdk.TrackierSDK.setUserId(optString);
        return true;
    }

    private boolean setUserEmail(String optString) {
        com.trackier.sdk.TrackierSDK.setUserEmail(optString);
        return true;
    }

    private boolean setUserName(String optString) {
        com.trackier.sdk.TrackierSDK.setUserName(optString);
        return true;
    }

    private boolean setUserPhone(String optString) {
        com.trackier.sdk.TrackierSDK.setUserPhone(optString);
        return true;
    }

    private boolean setDOB(String optString) {
        com.trackier.sdk.TrackierSDK.setDOB(optString);
        return true;
    }

    private boolean setGender(String optString) {
        switch (optString) {
            case "Male":
                com.trackier.sdk.TrackierSDK.setGender(TrackierSDK.Gender.MALE);
                break;
            case "Female":
                com.trackier.sdk.TrackierSDK.setGender(TrackierSDK.Gender.Female);
                break;
            case "Others":
                com.trackier.sdk.TrackierSDK.setGender(TrackierSDK.Gender.OTHERS);
                break;
            default:
                Log.d("TrackierSDK", "No Genders Found");
                break;
        }
        return true;
    }

    private boolean parseDeepLink(String optString) {
        Uri uri = Uri.parse(optString);
        com.trackier.sdk.TrackierSDK.parseDeepLink(uri);
        return true;
    }

    private boolean fireInstall() {
        com.trackier.sdk.TrackierSDK.fireInstall();
        return true;
    }

}
